main.go

```diff
@@ -84,35 +87,98 @@ type Server struct {
 	sync.Mutex // TODO: Do we really need a mutex?
 	raw        *C.RandomnessServer
 	noCopy     noCopy //nolint:structcheck
+	md         uint8
+	done       chan bool
+	pubKey     []byte
 }
 
-func serverFinalizer(server *Server) {
-	C.randomness_server_release(server.raw)
-	server.raw = nil
+// epochLoop periodically punctures the randomness server's PPOPRF and -- if
+// necessary -- re-creates the randomness server instance.
+func (srv *Server) epochLoop(epochLen time.Duration) {
+	ticker := time.NewTicker(epochLen)
+	elog.Println("Starting epoch loop.")
+	for {
+		select {
+		case <-srv.done:
+			return
+		case <-ticker.C:
+			if err := srv.puncture(); err != nil {
+				if err.Error() == errEpochExhausted {
+					if err := srv.init(); err != nil {
+						elog.Fatal("Failed to re-initialize randomness server.")
+					}
+				}
+			}
+		}
+	}
 }
 
-// NewServer returns a new PPOPRF randomness server instance.
-//
-// FIXME Pass in a list of 8-bit tags defining epochs.
-// The instance will generate its own secret key.
-func NewServer() (*Server, *string, error) {
+// init (re-)initializes the randomness server instance of the Rust FFI.
+func (srv *Server) init() error {
+	srv.Lock()
+	defer srv.Unlock()
+
 	// FIXME should we runtime.LockOSThread() here?
```

Comment:
I removed the FIXME in f605acf. As far as I can tell, we don't need `LockOSThread()` because sta-rs doesn't require thread local storage, and also doesn't require all calls to come from the same thread. According to [this stackoverflow comment](https://stackoverflow.com/a/25362395), `LockOSThread()` is useful for graphics libraries because they often do impose such restrictions.

That said, you are right, we are missing locks. The HTTP framework doesn't take care of that for us. I addressed the issue in 13eef5b.